File: C++/Cristian Maldonado/Ordenamientos/Bitonic_Sort.cpp
#include <iostream>

using namespace std;

void compAndSwap(int arr[], int i, int j, int dir){
    if((arr[i] > arr[j] && dir == 1)||(arr[i] < arr[j] && dir == 0)){
        swap(arr[i], arr[j]);
    }
}

void bitonicMerge(int arr[], int low, int cnt, int dir){
    if(cnt > 1){
        int k = cnt / 2;
        for(int i = low; i < low + k; i++){
            compAndSwap(arr, i, i + k, dir);
        }
        bitonicMerge(arr, low, k, dir);
        bitonicMerge(arr, low + k, k, dir);
    }
}

void bitonicSort(int arr[], int low, int cnt, int dir){
    if(cnt > 1){
        int k = cnt / 2;
        bitonicSort(arr, low, k, 1);
        bitonicSort(arr, low + k, k, 0);
        bitonicMerge(arr, low, cnt, dir);
    }
}

void sorted(int arr[], int N, int up){
    bitonicSort(arr, 0, N, up);
}

void printArray(int arr[], int n){
    for(int i = 0; i < n; ++i){
        if(i == n - 1){
            cout<<arr[i]<<endl;
            break;
        }
        cout << arr[i]<<" ";
    }
}

int main()
{
    int arr[] = {-5, -6, -1, -4, 8, 100, -90, -15, 35};
    int up = 1;
    int tam = sizeof(arr)/sizeof(arr[0]);
    sorted(arr, tam, up);
    cout<<"Array ordenado"<<endl;
    printArray(arr, tam);
    return 0;
}

File: Codigos estudiantes por lenguaje/PY/Julian Silva/algoritmos de ordenamiento/bitonicsort.py
from sys import stdin, stdout
rl = stdin.readline
wr = stdout.write


def compAndSwap(arr, i, j, direc):
    if (arr[i] > arr[j] and direc == 1) or (arr[i] < arr[j] and direc == 0):
        arr[i], arr[j] = arr[j], arr[i]


def bitonicMerge(arr, low, cnt, direc):

    if cnt > 1:
        k = cnt // 2
        for i in range(low, low+k):
            compAndSwap(arr, i, i+k, direc)
            
        bitonicMerge(arr, low, k, direc)
        bitonicMerge(arr, low+k, k, direc)


def bitonicSort(arr, low, cnt, direc):

    if cnt > 1:
        k = cnt // 2
        bitonicSort(arr, low, k, 1)
        bitonicSort(arr, low+k, k, 0)
        bitonicMerge(arr, low, cnt, direc)


def sortB(arr, n, up):
    bitonicSort(arr, 0, n, up)


def printArray(arr):
    for i in arr:
        wr(f'{i} ')
    wr('\n')


arr = list(map(int, rl().split()))
up = 1
printArray(arr)
sortB(arr, len(arr), up)
printArray(arr)